addons/mail/controllers/discuss.py

```diff
@@ -573,3 +573,29 @@ def mail_guest_update_name(self, guest_id, name):
         if guest_to_rename_sudo != guest and not request.env.user._is_admin():
             raise NotFound()
         guest_to_rename_sudo._update_name(name)
+
+    # --------------------------------------------------------------------------
+    # Link preview API
+    # --------------------------------------------------------------------------
+
+    @http.route('/mail/link_preview', methods=['POST'], type='json', auth='public')
+    def mail_link_preview(self, message_id, urls):
+        throttle = request.env['ir.config_parameter'].sudo().get_param('mail.link_preview_throttle', 99)
+        if int(throttle) == 0:
+            return [('clear',)]
+        link_previews = []
+        guest = request.env['mail.guest']._get_guest_from_request(request)
+        for url in set(urls[0:5]):
+            link_preview = request.env['mail.link.preview'].sudo().with_context(guest=guest).create_link_preview(url, message_id)
+            if link_preview:
+                link_previews.append(('insert', link_preview._link_preview_format()))
+        return link_previews or [('clear',)]
+
+    @http.route('/mail/link_preview/delete', methods=['POST'], type='json', auth='public')
+    def mail_link_preview_delete(self, link_preview_id):
+        link_preview_sudo = request.env['mail.link.preview'].browse(int(link_preview_id)).sudo().exists()
+        guest = request.env['mail.guest']._get_guest_from_request(request)
+        if link_preview_sudo.message_id and not link_preview_sudo.message_id.is_current_user_or_guest_author and not guest.env.user._is_admin():
+            raise NotFound()
+        if link_preview_sudo:
+            link_preview_sudo._delete_and_notify()
```

Comment:
`_delete_and_notify() ` should be coded in a way that is fine to call on any recordsed (including the empty one)